demos/portal-embed-demo/public/js/portalog/portals-controller.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * Portals Controller
+ * Controlling all the portal related features
+ */
+class PortalsController {
+
+    /**
+     * Initiating paratmeter
+     * @constructor
+     * @param {HTMLElement} root - the root element
+     * @param {String} origin - origin of portal
+     * @param {String} path - path of portal
+     * @param {Number} anmiateY - A target position to animate in px
+     */
+    constructor(root, origin, path, anmiateY) {
+        this.root = root;
+        this.origin = origin;
+        this.path = path;
+        this.anmiateY = anmiateY;
+    }
+
+    /**
+     * Embedding portal element to the root.
+     * @param {HTMLPortalElement} predecessor - optional
+     */
+    embed(predecessor) {
+
+        // clean up the root
+        this._remove(this.root.querySelector('portal'));
+
+        if (predecessor) {
+            // If the optional predecessor parameter was set
+            // add the predecessor to the root
+            this.portal = predecessor;
+        } else {
+            // Or else, create new portal
```

Comment:
Having this handle two paths, even though they both only have one call site, seems a little confusing. I would suggest that instead, we structure the code as two separate cases. All that we need to do is replace the old portal with the new portal, hook up any event listeners (or we can avoid that by just registering the event listener on the parent node thanks to propagation), and restore the player UI.

```javascript
/**
 * Creates a new <portal> element and player UI.
 */
populateEmbedContainer() {
    this.portal = document.createElement('portal');
    this.portal.src = this.src;
    this.playerUI = this._genPortalMessageUI();
    this.embedContainer.append(this.portal, this.playerUI);

    // When the portal is clicked, start animating it.
    this.embedContainer.addEventListener('click', evt => {
        if (evt.target === this.portal)
            this.animateAndActivate();
    });
}

/**
 * Animates the embedded content to smoothly activate it.
 */
animateAndActivate() {
    this.playerUI.style.display = 'none';
    // (current logic to start the CSS transition)
}

/**
 * Reinstalls the embedded portal after returning to PORTALOG.
 * @param {HTMLPortalElement} predecessor A portal element containing the embedded page.
 */
returnFromEmbed(predecessor) {
    this.playerUI.style.display = '';
    this.portal.replaceWith(predecessor);
    this.portal = predecessor;
}
```

Reply:
Agreed. This makes things a lot easier to read! Will keep things minimum with scoping the concerns of each function.